src/gpl/src/mbff.cpp

```diff
@@ -0,0 +1,1058 @@
+#include "mbff.h"
+#include <iostream>
+#include <algorithm>
+#include <vector>
+#include <map>
+#include <set>
+#include <ctime>
+#include <chrono>
+#include <memory>
+#include <random>
+#include <omp.h>
+#include <ortools/linear_solver/linear_solver.h>
+#include <ortools/base/logging.h>
+#include <ortools/sat/cp_model.h>
+#include <ortools/sat/cp_model.pb.h>
+#include <ortools/sat/cp_model_solver.h>
+#include <lemon/list_graph.h>
+#include <lemon/network_simplex.h>
+#include <lemon/maps.h>
+
+/* 
+height and width of a flip-flop
+ratios are used for:
+    (1) finding all slot locations in a tray
+    (2) finding power consumption per slot 
+*/
+
+float WIDTH = 2.448, HEIGHT = 1.2;
+float RATIOS[6] = { 1, 0.875, 0.854, 0.854, 0.844, 0.844 };
+
+namespace gpl {
+
+int MBFF::GetBitCnt(int bit_idx) { return (1 << bit_idx); }
+
+float MBFF::RunLP(std::vector<Flop> flops, std::vector<Tray> &trays,
+                  std::vector<std::pair<int, int> > clusters, int sz) {
+
+    int num_flops = static_cast<int>(flops.size()), num_trays = static_cast<int>(trays.size());
+
+    std::unique_ptr<operations_research::MPSolver> solver(
+        operations_research::MPSolver::CreateSolver("GLOP"));
+    const double inf = solver->infinity();
+
+    std::vector<const operations_research::MPVariable *> tray_x(num_trays),
+        tray_y(num_trays);
+    for (int i = 0; i < num_trays; i++) {
+        tray_x[i] = solver->MakeNumVar(0, inf, "");
+        tray_y[i] = solver->MakeNumVar(0, inf, "");
+    }
+
+    std::vector<const operations_research::MPVariable *> d_x(num_flops),
+        d_y(num_flops);
+
+    for (int i = 0; i < num_flops; i++) {
+
+        d_x[i] = solver->MakeNumVar(0, inf, "");
+        d_y[i] = solver->MakeNumVar(0, inf, "");
+
+        int tray_idx = clusters[i].first, slot_idx = clusters[i].second;
+
+        Point flop = flops[i].pt;
+        Point tray = trays[tray_idx].pt;
+        Point slot = trays[tray_idx].slots[slot_idx];
```

Comment:
Perhaps `const Point&`? I think you are not changing these variables below, so you don't need to copy it.